File: src/date_pattern.py
# get_dates must be inclusive of the end and exclusive of the start
class UnionDatePattern:
    def __init__(self, date_patterns):
        self.date_patterns = date_patterns

    def to_dict(self):
        return {
            'datePatterns': self.date_patterns
        }

    @staticmethod
    def from_dict(data):
        return UnionDatePattern(data['datePatterns'])

    def get_dates(self, start, end):
        return [date for p in self.date_patterns for date in p.get_dates(start, end)]

class UniformDatePattern:
    def __init__(self, initial_time, delta_time):
        self.initial_time = initial_time
        self.delta_time = delta_time

    def to_dict(self):
        return {
            'initialTime': self.initial_time,
            'deltaTime': self.delta_time
        }

    @staticmethod
    def from_dict(data):
        return UniformDatePattern(data['initialTime'], data['deltaTime'])

    def get_dates(self, start, end):
        if initial_time > start:
            time = initial_time
        else:
            n = (start - initial_time) // delta_time + 1
            time = initial_time + n * delta_time
        result = []
        while time <= end:
            result.append(time)
            time += delta_time
        return result

# TODO
class MonthlyDatePattern:
    def __init__(self):
        pass

    def to_dict(self):
        pass

    @staticmethod
    def from_dict(data):
        pass

    def get_dates(self, start, end):
        pass

File: src/main/java/io/github/theimbichner/taskmanager/time/DatePattern.java
package io.github.theimbichner.taskmanager.time;

import java.time.Instant;

import io.vavr.collection.Vector;

import org.json.JSONException;
import org.json.JSONObject;

public interface DatePattern {
   static DatePattern fromJson(JSONObject json) {
      String patternType = json.getString("patternType");
      switch(json.getString("patternType")) {
      case "union":
         return UnionDatePattern.fromJson(json);
      case "uniform":
         return UniformDatePattern.fromJson(json);
      case "monthly":
         return MonthlyDatePattern.fromJson(json);
      default:
         String format = "Unrecognized patternType: %s";
         String message = String.format(format, patternType);
         throw new JSONException(message);
      }
   }

   /**
    * Returns a list of matching Instants that lie between start (exclusive) and
    * end (inclusive). Start must be strictly before end. The resulting list
    * will be sorted.
    *
    * @param start The start of the time range, exclusive
    * @param end The end of the time range, inclusive
    * @return A list of matching Instants between start and end.
    */
   Vector<Instant> getDates(Instant start, Instant end);
   JSONObject toJson();
}
